### PR TITLE
Improve server selection ordering

### PR DESCRIPTION
**Pull request by pglass**
This fixes poor behavior where the Watcher could (at random) potentially reconnect to servers it just disconnected from, without trying all other servers. Previously, whenever it connected to the server watch stream it would "reset" its knowledge of server addresses, which means it would pick the next server address at random.

To fix this, the Watcher now attempts to connect to all known addresses once, before trying the same server twice.

The Watcher tracks the following information for each server address:

1. lastAttempt: the last time it attempted to each server address.
2. lastUpdate: the last time an update about the address was received from either discovery or from the server watch stream.

When selecting the next server address, the Watcher chooses an address that it has not attempted since the last update, and otherwise chooses the least-recently attempted address if there are multiple. Because it sorts by least-recently attempted addresses, it will rotate through all servers in its current list before trying the same server twice.

### CODE REVIEW

**Review by boxofrad — APPROVED** [review]
Great job finding and fixing this! :clap:

I wondered whether not hanging onto an old server's `lastAttempt` timestamp when it disappears from the stream could cause problems when a server's health is flapping. We do [handle Raft voters falling behind a little](https://github.com/hashicorp/consul/blob/e8fcbfe19e5de071cbfecdeed33ee75d37ffb447/agent/consul/autopilotevents/ready_servers_events.go#L217) to reduce flapping, but not for read-replicas or servers failing gossip checks (e.g. due to a network partition).

On reflection, I couldn't think of a scenario where this would be particularly problematic that wasn't also quite contrived. So I think it's better to avoid extra complexity.

**tl;dr:** LGTM

File: discovery/watcher.go [hunk 1]
```diff
@@ -488,85 +474,82 @@ func (w *Watcher) getDataplaneFeatures() (map[string]bool, error) {
...
 				w.log.Warn("failed to parse server address from watch stream", "error", err)
 				continue
 			}
-			set.Put(OK, addr)
+			streamAddrs = append(streamAddrs, addr)
 		}
+
+		// Update the addrSet. This sets the lastUpdated timestamp for each address,
+		// and removes servers not present in the server watch stream.
+		addrs.SetAddrs(streamAddrs...)
```
[the remark this answers is not in the document]
Comment (boxofrad) [reply discovery/watcher.go:532]: Yeah that's true, but I think in that case the offending server wouldn't consider itself unhealthy and push an update on the stream to remove itself - so when the clients eventually figure out that something is wrong (e.g. via missed keepalive ACKs) they'd just move on to the next server.

I guess it could still be a problem if whatever server they land on is already oversubscribed, though?

I'm also wondering whether there's a potential failure-mode where the server you're connected to is so overwhelmed it's unable to process memberlist and autopilot traffic, so tells you that it is the only healthy server. I guess in that case, falling back to DNS/script resolution would fix it - sorry, just thinking aloud!

[the remark this answers is not in the document]
Comment (pglass) [reply discovery/watcher.go:532]: I merged this, but making sure to conclude this discussion.

Basically, my thought is probably not a big concern because:

1. consul-dataplanes would only switch servers when they are disconnected, and servers flapping in the server watch stream would not immediately cause them to be disconnected
2. consul-dataplane should not forget about the server it is currently connected to, because that server would not flap in the server watch stream, because a server would not consider itself unhealthy.
3. consul-dataplane will try _different_ server on its next attempt, because it should always remember it attempted the server it was most-recently connected to, because of (2), even if it has forgotten all other servers
